src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        [ThreadStatic] private static DisposeScopeManager _threadSingleton;
+        internal ThreadDisposeScopeStatistics StatisticsInstance { get; } = new ThreadDisposeScopeStatistics();
```

Comment:
NIT: why not just `Statistics` and `DisposeScopeStatistics` for class name?

Reply:
Ah, because the statistics apply to the thread you're running in, a different thread will have different statistics. To make that clear(er) that's the naming we arrived at. I'd happily change to something better, would StatisticsForThread be more explanatory?